validator/coordinator/src/test/java/tech/pegasys/artemis/validator/coordinator/DepositProviderTest.java

```diff
@@ -0,0 +1,175 @@
+import tech.pegasys.artemis.datastructures.state.Checkpoint;
+import tech.pegasys.artemis.datastructures.util.DataStructureUtil;
+import tech.pegasys.artemis.datastructures.util.DepositUtil;
+import tech.pegasys.artemis.datastructures.util.MerkleTree;
+import tech.pegasys.artemis.datastructures.util.OptimizedMerkleTree;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+import tech.pegasys.artemis.storage.events.FinalizedCheckpointEvent;
+import tech.pegasys.artemis.util.SSZTypes.SSZList;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class DepositProviderTest {
+
+  private int seed = 0;
+
+  private List<tech.pegasys.artemis.pow.event.Deposit> allSeenDepositsList;
+  private DepositProvider depositProvider;
+  private ChainStorageClient chainStorageClient;
+  private BeaconState beaconState;
+  private MerkleTree depositMerkleTree;
+
+  @BeforeEach
+  void setUp() {
+    chainStorageClient = mock(ChainStorageClient.class);
+    beaconState = mock(BeaconState.class);
+
+    depositMerkleTree = new OptimizedMerkleTree(Constants.DEPOSIT_CONTRACT_TREE_DEPTH);
+    depositProvider = new DepositProvider(chainStorageClient);
+
+    createDepositEvents(40);
+  }
+
+  @Test
+  void stateEth1DepositIndexIsEqualToEth1DataDepositCount_NoDepositReturned() {
+    Constants.MAX_DEPOSITS = 5;
+    mockStateEth1DepositIndex(2);
+    mockEth1DataDepositCount(2);
+    mockDepositsFromEth1Block(0, 10);
+    SSZList<Deposit> deposits = depositProvider.getDeposits(beaconState);
+    assertThat(deposits).isEmpty();
+  }
+
+  @Test
+  void numberOfDepositsThatCanBeIncludedLessThanMaxDeposits() {
+    mockStateEth1DepositIndex(5);
+    mockEth1DataDepositCount(20);
+
+    Constants.MAX_DEPOSITS = 16;
+
+    mockDepositsFromEth1Block(0, 10);
+    mockDepositsFromEth1Block(10, 20);
+
+    SSZList<Deposit> deposits = depositProvider.getDeposits(beaconState);
+    assertThat(deposits).hasSize(15);
+    checkThatDepositProofIsValid(deposits);
+  }
+
+  @Test
+  void numberOfDepositsThatCanBeIncludedMoreThanMaxDeposits() {
+    mockStateEth1DepositIndex(5);
+    mockEth1DataDepositCount(20);
+
+    Constants.MAX_DEPOSITS = 10;
+
+    mockDepositsFromEth1Block(0, 10);
+    mockDepositsFromEth1Block(10, 20);
+
+    SSZList<Deposit> deposits = depositProvider.getDeposits(beaconState);
+    assertThat(deposits).hasSize(10);
+    checkThatDepositProofIsValid(deposits);
+  }
+
+  @Test
+  void depositsWithFinalizedIndicesGetPrunedFromMap() {
+    Bytes32 finalizedBlockRoot = Bytes32.fromHexString("0x01");
+    mockStateEth1DepositIndex(10);
+    mockDepositsFromEth1Block(0, 20);
+    when(chainStorageClient.getBlockState(eq(finalizedBlockRoot)))
+        .thenReturn(Optional.ofNullable(beaconState));
+
+    depositProvider.onFinalizedCheckpoint(
+        new FinalizedCheckpointEvent(new Checkpoint(UnsignedLong.ONE, finalizedBlockRoot)));
+
+    assertThat(depositProvider.depositNavigableMap)
```

Comment:
Rather than expose this map, I'd add a `size` or `depositCount` method to return how many deposits are still stored, then check that is reduced after a finalised block and that the expected remaining deposits are still returned on subsequent calls to `getDeposits`.  That way we're testing the expected behaviours of the API rather than internal details of the implementation.  While we don't strictly need a `size` method, it's a sensible addition to the API which could be used by monitoring and is better than exposing the internal map directly.

Reply:
Makes sense. done.